### PR TITLE
Implementing server for OAuth2 protocol

### PR DESCRIPTION
**Pull request by heyaudace**
## Quick notes:
* Server is instantiated by by passing in a handler, a delegate that will be invoked on the authorization response(auth_code or error) when the response becomes available.
* Server picks an endpoint at random that it will be listening from.
* The endpoint serves as redirect_uri for receiving authorization code.
* This change does not affect the UI.

### CODE REVIEW

**Comment by heyaudace** [comment]
## PR Ready for review.

* Re-did server implementation to make it simpler with more concise code.
* Wrote tests for the server.
* Improved documentation as well.

**Comment by heyaudace** [comment]
### Adding screenshots for authorization page and redirect endpoint:
* Authorization page:
screen shot 2018-08-06 at 2 12 44 pm

* Redirect endpoint:
screen shot 2018-08-06 at 2 07 50 pm

**Comment by heyaudace** [comment]
### Changed implementation and tests according to today's discussion. 
PR good to go!

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+            return result;
+        }
+
+        /// <summary>
+        /// Stops the server. A future call of the Start() method on this instance will restart this server at a
+        /// different endpoint.
+        /// </summary>
+        private void Stop()
+        {
+            _httpListener.Close();
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Wrap in null check so no crash if called 2+ times?

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+        /// <summary>
+        /// Processes URI, extracts query params, puts them into a dictionary returns the dictionary.
+        /// </summary>
+        internal static Dictionary<string, string> GetQueryParamsFromUri(Uri uri)
+        {
+            var result = new Dictionary<string, string>();
+            foreach (var pair in uri.Query.Substring(1).Split('&'))
+            {
+                if (pair.Contains("="))
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: When would this be false?

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+                   noOtherKeysArePresent(queryParams);
+        }
+
+        /// <summary>
+        /// Processes URI, extracts query params, puts them into a dictionary returns the dictionary.
+        /// </summary>
+        internal static Dictionary<string, string> GetQueryParamsFromUri(Uri uri)
+        {
+            var result = new Dictionary<string, string>();
+            foreach (var pair in uri.Query.Substring(1).Split('&'))
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Extract a variable and add a comment above that line:

```
// Uri's Query string always starts with "?" so skip the first character.
var query = uri.Query.Substring(1);
foreach (var pair in query.Split('&'))
```

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    /// request, which it will process to retrieve the authorization response and handle the by invoking on the response
+    /// the handler passed to the server during instatiation. The server will then stop listening for further requests.
+    ///
+    /// <see cref="https://developers.google.com/identity/protocols/OAuth2#installed"/> for an overview of OAuth2
+    /// protocol for installed applications that interact with Google APIs.
+    /// </summary>
+    public class OAuth2Server
+    {
+        private const string CloseTabText = "You may close this tab.";
+        private const int ServerPort = 2806;
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Nit: insert line below to separate const fields from instance fields.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        private HttpListener _httpListener;
+        private string _callbackEndpoint;
+
+        private readonly Action<KeyValuePair<string, string>> _onResponseAction;
+
+        /// <summary>
+        /// An endpoint that on which the server is listening.
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Nit: remove "that"

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+        /// <summary>
+        /// An instance of a server that will run locally to retrieve authorization code. The server will stop running
+        /// once the first response gets received.
+        /// </summary>
+        /// <param name="onResponseAction">An action to be invoked on the key-value pair representing the first
+        /// response that will be caught by the server. Note that that the invocation of this handler will not be done
+        /// on the main thread, therefore the handler should only run operations that can be run on the main thread.
+        /// </param>
+        public OAuth2Server(Action<KeyValuePair<string, string>> onResponseAction)
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Nit: move constructor to line 49 (after fields but before first method/property)

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+        /// <summary>
+        /// Starts this server to make it listen for requests containing authorization code or error data that are
+        /// forwarded from google's OAuth2 authorization url.
+        ///
+        /// After a call to this method, the CallBackEndpoint property of this instance will provide the endpoint
+        /// at which this server is listening.
+        /// </summary>
+        public void Start()
+        {
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Is it OK for Start() to be called more than once? If not, check _callbackEndpoint for not null and throw an exception.

If it is OK, check for _httpListener being null (indicating never started or Stop() called) and throw an exception if necessary.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+            _callbackEndpoint = string.Format("http://localhost:{0}/{1}/", ServerPort, Path.GetRandomFileName());
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Depending on the answer to the question above, this can be readonly and set in the constructor.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+    /// its chosen endpoint with the CallbackEndpoint property. This server will run until it receives the first
+    /// request, which it will process to retrieve the authorization response and handle the by invoking on the response
+    /// the handler passed to the server during instatiation. The server will then stop listening for further requests.
+    ///
+    /// <see cref="https://developers.google.com/identity/protocols/OAuth2#installed"/> for an overview of OAuth2
+    /// protocol for installed applications that interact with Google APIs.
+    /// </summary>
+    public class OAuth2Server
+    {
+        private const string CloseTabText = "You may close this tab.";
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs:40]: Add a comment:

// Arbitrarily chosen port to listen for the authorization callback.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+                codeOrErrorKeyMustBePresentText);
+
+
+            const string codeAndErrorCannotBePresentText = "\"code\" and \"error\" cannot be present at the same time.";
+            var invalidUriWithCodeAndError =
+                new Uri(string.Format("{0}?code=codeValue&error=errorValue", AddressPrefix));
+            Assert.IsFalse(OAuth2Server.UriContainsValidQueryParams(invalidUriWithCodeAndError),
+                codeAndErrorCannotBePresentText);
+
+            const string noOtherKeysCanBePresentText =
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Nit: inline this

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        /// </summary>
+        internal static Dictionary<string, string> GetQueryParamsFromUri(Uri uri)
+        {
+            var result = new Dictionary<string, string>();
+            foreach (var pair in uri.Query.Substring(1).Split('&'))
+            {
+                if (pair.Contains("="))
+                {
+                    var keyAndValue = pair.Split('=');
+                    result.Add(keyAndValue[0], keyAndValue[1]);
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Do you need to unescape these strings?
https://docs.unity3d.com/ScriptReference/WWW.UnEscapeURL.html

Potentially not if the Google auth API guarantees that content is already URL safe. If this is the case, add a comment saying so.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            var invalidUriWithOtherKeys =
+                new Uri(string.Format("{0}?code=codeValue&otherKey=someValue", AddressPrefix));
+            Assert.IsFalse(OAuth2Server.UriContainsValidQueryParams(invalidUriWithOtherKeys),
+                noOtherKeysCanBePresentText);
+        }
+
+        [Test]
+        public void TestGetQueryParamsFromUri()
+        {
+            var testDict = new Dictionary<string, string>();
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Please change to something simpler like this:

var expected = new Dictionary<string, string> {{"a", "b"}, {"c", "d"}};
var uri = new Uri(AddressPrefix + "?a=b&c=d");

Also test boundary conditions like null and empty

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    public class OAuth2ServerTest
+    {
+        // Testing strategy:
+        //     - Test methods that do not require starting the server.
+        //     - TODO: Implement E2E tests to ensure that the server handles requests as expected.
+
+        private const string AddressPrefix = "http://localhost:5000/";
+
+        [Test]
+        public void TestGetAuthorizationResponse()
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs:36]: Also add a test for this case:

`OAuth2Server.GetAuthorizationResponse(new Uri(AddressPrefix));`

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+            // respose is invalid
+            var invalidUrl = new Uri(string.Format("{0}?someKey=someValue", AddressPrefix));
+            Assert.Throws<ArgumentException>(() => OAuth2Server.GetAuthorizationResponse(invalidUrl),
+                "Expected ArgumentException");
+            LogAssert.Expect(LogType.Error, "Uri query contains invalid params");
+        }
+
+        [Test]
+        public void TestUriContainsValidParams()
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Also add a test for this case:

`OAuth2Server.UriContainsValidQueryParams(new Uri(AddressPrefix));`

**Review by thanwah — CHANGES_REQUESTED** [review]
Thanks, this is looking simpler. Requesting a few changes.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 7 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Made _callbackEndpoint a readonly that is set after during initialization. Made it okay for Start() to be called more than once, but also made Stop pause listening for requests instead of permanently closing the httplistener.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 8 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Sure!

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 2 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Resolved.
 It was me trying to be a little defensive with the thought that '&' could be present in a non-query string, but now that I think about it it doesn't make sense, assuming Uri.Query does all the parsing and returns a valid query.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 11 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: The google API docs do not say anything about it, but the general assumption is that the authorization code does not include any special characters. Makes sense to unescape though since there are no explicit guarantees.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 1 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Made httpListener readonly and calling Stop() instead of Close() to support starting the second time without re-instantiating _httpListener, and this removed the need for null check.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 13 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs:36]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 14 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 10 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 12 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

Handled empty boundary case. Null doesn't make sense here since Uri.Query returns empty string when there are no query params in the Url string. Also made it explicit on the GetQueryParamsFromUri method that the uri passed itself  should not be null.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 3 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 6 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 5 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 9 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs:40]: Resolved.

**Review by heyaudace — COMMENTED** [review]
## Resolved all comments.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+            var errorResponse =
+                OAuth2Server.GetAuthorizationResponse(new Uri(string.Format("{0}?error={1}", AddressPrefix,
+                    someString)));
+            Assert.AreEqual(new KeyValuePair<string, string>("error", someString), errorResponse,
+                "Expected invalid error response");
+
+            // response has invalid keys
+            var invalidResponse = new Uri(string.Format("{0}?someKey=someValue", AddressPrefix));
+            Assert.Throws<ArgumentException>(() => OAuth2Server.GetAuthorizationResponse(invalidResponse),
+                "Expected ArgumentException");
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Delete  "Expected ArgumentException" param since this is very clear from context.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+            LogAssert.Expect(LogType.Error, OAuth2Server.InvalidQueryExceptionMessage);
+
+            // No response. Uri has no params
+            Assert.Throws<ArgumentException>(() => OAuth2Server.GetAuthorizationResponse(new Uri(AddressPrefix)),
+                "Expected ARgumentException");
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Same.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        [Test]
+        public void TestGetQueryParamsFromUri()
+        {
+            // Case 1: Non-empty query
+            var testDictionary = new Dictionary<string, string> {{"a", "b"}, {"c", "d"}};
+            var uri = new Uri(AddressPrefix + "?a=b&c=d");
+            Assert.AreEqual(testDictionary, OAuth2Server.GetQueryParamsFromUri(uri),
+                "Should return expected correct query params");
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Same. Delete since clear from context

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+            var invalidUriWithEmptyQuery = new Uri(AddressPrefix);
+            Assert.IsFalse(OAuth2Server.UriContainsValidQueryParams(invalidUriWithEmptyQuery),
+                "Uri with empty query should be invalid");
+        }
+
+        [Test]
+        public void TestGetQueryParamsFromUri()
+        {
+            // Case 1: Non-empty query
+            var testDictionary = new Dictionary<string, string> {{"a", "b"}, {"c", "d"}};
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Rename to expected or expectedDictionary.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+                queryParamsDict.Where(kvp => !allowedQueries.Contains(kvp.Key)).ToArray().Length == 0;
+            return codeOrErrorIsPresent(queryParams) && notBothCodeAndErrorArePresent(queryParams) &&
+                   noOtherKeysArePresent(queryParams);
+        }
+
+        /// <summary>
+        /// Processes Uri, extracts query params, puts them into a dictionary returns the dictionary.
+        /// Uri must not be null.
+        /// </summary>
+        internal static Dictionary<string, string> GetQueryParamsFromUri(Uri uri)
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Delete method (see above).

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+
+        /// <summary>
+        /// Returns a key value pair corresponding to the authorization response sent from OAuth2 authorization page.
+        /// Logs error message and throws ArgumentException if the uri contains invalid params.
+        /// </summary>
+        /// <param name="uri">The uri of the incoming request</param>
+        /// <exception cref="ArgumentException">Exception thrown if the uri contains invalid params.</exception>
+        internal static KeyValuePair<string, string> GetAuthorizationResponse(Uri uri)
+        {
+            if (!UriContainsValidQueryParams(uri))
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: To be honest, I was a bit confused about what exactly the code was trying to do until the latest PR that simplifies things a bit. The current approach relies on calling GetQueryParamsFromUri() twice which is a bit inefficient. 

So I'm going to suggest something that is hopefully simpler and that provides an error message that specifically identifies the issue as opposed to a generic message that "Uri query is not valid".

```
internal static KeyValuePair<string, string> GetAuthorizationResponse(Uri uri)
{
  var query = GetQueryString(uri);
  return GetCodeOrErrorResponsePair(query);
}

internal static string GetQueryString(uri)
{
var query = uri.Query;
if (string.IsNullOrEmpty(query))
{
  throw new ArgumentException("URI is missing query parameters");
}
if (query[0] != "?")
{
  throw new ArgumentException("Expect first character of URI query is \"?\"");
}
return query.Substring(1);
}

internal static KeyValuePair<string, string> GetCodeOrErrorResponsePair(string query)
{
// The authorization response should consist of a single query parameter.
var paramPairs = query.Split('&');
if (paramPairs.Length != 1)
{
  throw new ArgumentException(string.Format("Unexpected number of URI parameters {0}", paramPairs.Length));
}

// The single query parameter should be an equals-separated key/value pair.
var keyAndValue = paramPairs[0].Split('=');
if (keyAndValue.Length != 2)
{
  throw new ArgumentException(string.Format("URI parameter pair has {0} components", keyAndValue.Length));
}
// The key for the single query parameter should either be "code" or "error".
var key = keyAndValue[0];
var value = Uri.UnescapeDataString(keyAndValue[1];
if (key != "code" || key != "error")
{
  throw new ArgumentException(string.Format("Unexpected URI parameter key \"{0}\"", key));
}
return new KeyValuePair<string, string>() {key, value}; // Not sure about format here, whatever is correct/concise.
}
```

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+            return GetQueryParamsFromUri(uri).ToArray()[0];
+        }
+
+        /// <summary>
+        /// Inspect the URI and determine whether it contains valid params according to the following policies:
+        ///   1. URI query must include "code" or "error" in param keys.
+        ///   2. "code" and "error" cannot be present at the same time.
+        ///   3. No other keys apart from "code", "error" are allowed.
+        /// </summary>
+        internal static bool UriContainsValidQueryParams(Uri uri)
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Delete method (see above).

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+using UnityEngine;
+using UnityEngine.TestTools;
+
+namespace GooglePlayInstant.Tests.Editor.QuickDeploy
+{
+    /// <summary>
+    /// Contains unit tests for OAuth2Server methods.
+    /// </summary>
+    [TestFixture]
+    public class OAuth2ServerTest
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs:27]: After the change I requested, you'll have to rearrange the tests a bit. However, mostly the tests should be the same.

There should be one new test though that checks for Uri.UnescapeDataString() working correctly, e.g. " " to "%20" as in the other PR.

**Review by thanwah — COMMENTED** [review]
Thanks for making the changes from last time. I'm sorry but I don't want to submit just yet because I thought of a way to make the code a little clearer and handle more edge cases. Please take a look.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 20 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 21 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 19 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 22 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs:27]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 15 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 16 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 18 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 17 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

**Review by heyaudace — COMMENTED** [review]
## Made necessary changes on implementation and tests according to review comments.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+        internal static KeyValuePair<string, string> GetAuthorizationResponse(Uri uri)
+        {
+            var query = GetQueryString(uri);
+            return GetCodeOrErrorResponsePair(query);
+        }
+        
+        /// <summary>
+        /// Returns a query string from the given uri.
+        /// </summary>
+        /// <param name="uri"></param>
```
Comment (thanwah) [inline GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Delete line

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+        // Testing strategy:
+        //     - Test methods that do not require starting the server.
+        //     - TODO: Implement E2E tests to ensure that the server handles requests as expected.
+
+        private const string AddressPrefix = "http://localhost:5000/";
+
+        [Test]
+        public void TestEscapeAndUnescapeDataString()
+        {
+            Assert.AreEqual("%20", Uri.EscapeDataString(" "));
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Sorry I wasn't clear. This is testing Uri.EscapeDataString() which isn't something that we need to test. The test I was looking for is to verify that you're calling Uri.UnescapeDataString() where appropriate in GetCodeOrErrorResponsePair(). You'd do this by passing in something like "code=%20" and verify that the dictionary response is {"code", " "}.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 25]
```diff
@@ -0,0 +1,109 @@
...
+            Assert.Throws<ArgumentException>(() => OAuth2Server.GetAuthorizationResponse(new Uri(AddressPrefix)));
+        }
+
+        [Test]
+        public void TestGetQueryString()
+        {
+            var validUriWithCode = new Uri(string.Format("{0}?code=someValue", AddressPrefix));
+            Assert.AreEqual("code=someValue", OAuth2Server.GetQueryString(validUriWithCode));
+
+            var validUriWithError = new Uri(string.Format("{0}?error=someValue", AddressPrefix));
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: GetQueryString doesn't care about the params. So this test is redundant and can be deleted.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+
+        [Test]
+        public void TestGetCodeOrErrorResponsePairOnInvalidInputs()
+        {
+            // The authorization response should consist of a single query parameter.
+            Assert.Throws<ArgumentException>(
+                () => OAuth2Server.GetCodeOrErrorResponsePair("code=codeValue&error=errorValue"));
+            Assert.Throws<ArgumentException>(
+                () => OAuth2Server.GetCodeOrErrorResponsePair("code=codeValue&otherKey=someValue"));
+            Assert.Throws<ArgumentException>(() =>
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Nit: please wrap consistently (like lines above)

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+                OAuth2Server.GetCodeOrErrorResponsePair("error=errorValue&otherKey=someValue"));
+
+            // The single query parameter should be an equals-separated key/value pair.
+            Assert.Throws<ArgumentException>(() => OAuth2Server.GetCodeOrErrorResponsePair("code&codeValue"));
+
+            // Key for the single query parameter should be either "code" or "error"
+            Assert.Throws<ArgumentException>(() => OAuth2Server.GetCodeOrErrorResponsePair("someKey=someValue"));
```
Comment (thanwah) [inline GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Some potential test cases to add:
""
"&"
"="
"code"
"code=value1=value2"

**Review by thanwah — APPROVED** [review]
LGTM. A few additional comments, but can be merged after resolving them.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 24 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 27 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 25 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs (hunk 26 above)
Comment (heyaudace) [reply GooglePlayInstant/Tests/Editor/QuickDeploy/OAuth2ServerTest.cs]: Resolved.

File: GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs (hunk 23 above)
Comment (heyaudace) [reply GooglePlayInstant/Editor/QuickDeploy/OAuth2Server.cs]: Resolved.

**Review by heyaudace — COMMENTED** [review]
## Resolved final comments. Will merge
